File: src/rp_ylx/recording/stereo_encoder.py
"""驱动板上 `ylx-stereo-encoder` 助手进程的边录边编码接口。

录制守护进程继续拥有采集、帧序、IMU 和丢帧记账；本模块只负责把每一帧并排
MJPEG 交给助手进程，并把助手封好的分段回报给录制器。助手独立成进程的原因是
JPU/VPU 卡死时只应让本次录制失败，而不是拖垮守护进程。
"""

from __future__ import annotations

import json
import os
import shutil
import struct
import subprocess
import sys
import threading
from collections.abc import Sequence
from dataclasses import dataclass
from pathlib import Path

from rp_ylx.native import NativeModuleError, NativeSessionIo, create_native_session_io

_FRAME_MAGIC = b"YLXF"
_HEADER = struct.Struct("<4sI")
_READY_TIMEOUT_SECONDS = 15.0
_DEFAULT_EXECUTABLE = "ylx-stereo-encoder"
_SESSION_IO_LOCK = threading.Lock()
_SESSION_IO: NativeSessionIo | None = None
_SESSION_IO_UNAVAILABLE = False


class StereoEncoderError(RuntimeError):
    def __init__(self, code: str, message: str) -> None:
        self.code = code
        self.message = message
        super().__init__(f"{code}: {message}")


@dataclass(frozen=True, slots=True)
class ClosedSegment:
    """助手已经封好、可独立校验的一组左右眼分段。"""

    index: int
    start_frame: int
    end_frame: int
    left_path: str
    left_bytes: int
    right_path: str
    right_bytes: int


def resolve_executable(explicit: str | Path | None = None) -> Path:
    """定位助手可执行文件；找不到即视为本设备不具备边录边编码能力。"""

    if explicit is not None:
        candidate = Path(explicit)
        if candidate.is_file():
            return candidate
        raise StereoEncoderError("encoder_unavailable", f"助手不存在：{candidate}")
    override = os.environ.get("RP_YLX_STEREO_ENCODER")
    if override:
        return resolve_executable(override)
    # systemd 不会把发行版的 bin 目录放进 PATH，所以先看解释器旁边和
    # package-owned runtime 所属 release 的 bin 目录。
    executable = Path(sys.executable).resolve()
    beside = executable.parent / _DEFAULT_EXECUTABLE
    if beside.is_file():
        return beside
    if executable.parent.name == "bin" and executable.parent.parent.name == "runtime":
        release_bin = executable.parent.parent.parent / "bin" / _DEFAULT_EXECUTABLE
        if release_bin.is_file():
            return release_bin
    for entry in os.environ.get("PYTHONPATH", "").split(os.pathsep):
        if not entry:
            continue
        release_bin = Path(entry).resolve().parent / "bin" / _DEFAULT_EXECUTABLE
        if release_bin.is_file():
            return release_bin
    located = shutil.which(_DEFAULT_EXECUTABLE)
    if located is None:
        raise StereoEncoderError("encoder_unavailable", "未安装 ylx-stereo-encoder")
    return Path(located)


class StereoEncoderProcess:
    """一次录制对应一个助手进程；不复用、不重启。"""

    def __init__(
        self,
        out_dir: str | Path,
        *,
        executable: str | Path | None = None,
        width: int,
        height: int,
        fps: int,
        bitrate_kbps: int = 8192,
        segment_frames: int = 900,
        path_prefix: str = "video/",
    ) -> None:
        if width <= 0 or width % 4 or height <= 0 or fps <= 0 or segment_frames <= 0:
            raise ValueError("边录边编码参数无效")
        self._executable = resolve_executable(executable)
        self._out_dir = Path(out_dir)
        self._width = width
        self._height = height
        self._fps = fps
        self._bitrate_kbps = bitrate_kbps
        self._segment_frames = segment_frames
        self._path_prefix = path_prefix
        self._process: subprocess.Popen[bytes] | None = None
        self._reader: threading.Thread | None = None
        self._lock = threading.Lock()
        self._segments: list[ClosedSegment] = []
        self._failure: StereoEncoderError | None = None
        self._ready = threading.Event()
        self._done = threading.Event()
        self._stats: dict[str, int] = {}
        self._submitted = 0

    @property
    def segment_frames(self) -> int:
        return self._segment_frames

    @property
    def segments(self) -> tuple[ClosedSegment, ...]:
        with self._lock:
            return tuple(self._segments)

    @property
    def stats(self) -> dict[str, int]:
        with self._lock:
            return dict(self._stats)

    @property
    def submitted_frames(self) -> int:
        return self._submitted

    def start(self) -> None:
        if self._process is not None:
            raise StereoEncoderError("invalid_state", "助手进程只能启动一次")
        command = [
            str(self._executable),
            "--out-dir",
            str(self._out_dir),
            "--path-prefix",
            self._path_prefix,
            "--width",
            str(self._width),
            "--height",
            str(self._height),
            "--fps",
            str(self._fps),
            "--bitrate-kbps",
            str(self._bitrate_kbps),
            "--segment-frames",
            str(self._segment_frames),
        ]
        try:
            self._process = subprocess.Popen(  # noqa: S603 - 固定参数，无 shell
                command,
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.DEVNULL,
                bufsize=0,
            )
        except OSError as error:
            raise StereoEncoderError("encoder_unavailable", f"无法启动助手：{error}") from error
        self._reader = threading.Thread(
            target=self._read_events, name="rp-ylx-stereo-encoder-events", daemon=True
        )
        self._reader.start()
        if not self._ready.wait(_READY_TIMEOUT_SECONDS):
            self.abort()
            raise StereoEncoderError("encoder_unavailable", "助手未在期限内就绪")
        self._raise_if_failed()

    def submit(self, jpeg: bytes) -> None:
        """提交一帧并排 MJPEG。助手拒绝即视为本次录制失败，不做静默丢帧。"""

        process = self._process
        if process is None or process.stdin is None:
            raise StereoEncoderError("invalid_state", "助手进程未启动")
        self._raise_if_failed()
        try:
            native = _session_io_or_none()
            if native is None:
                _writev_all(process.stdin.fileno(), (_HEADER.pack(_FRAME_MAGIC, len(jpeg)), jpeg))
            else:
                written = native.write_encoder_frame(process.stdin.fileno(), jpeg)
                if written != _HEADER.size + len(jpeg):
                    raise BrokenPipeError("encoder pipe native write was short")
        except (BrokenPipeError, OSError, RuntimeError) as error:
            self._raise_if_failed()
            raise StereoEncoderError("encoder_failed", f"助手写入失败：{error}") from error
        self._submitted += 1

    def finish(self, *, timeout: float = 30.0) -> Sequence[ClosedSegment]:
        """关闭输入、等待助手封完最后一段，返回全部已封分段。"""

        process = self._process
        if process is None:
            raise StereoEncoderError("invalid_state", "助手进程未启动")
        if process.stdin is not None:
            try:
                process.stdin.write(_HEADER.pack(_FRAME_MAGIC, 0))
                process.stdin.flush()
                process.stdin.close()
            except OSError:
                pass
        try:
            code = process.wait(timeout=timeout)
        except subprocess.TimeoutExpired as error:
            self.abort()
            raise StereoEncoderError("encoder_failed", "助手未在期限内退出") from error
        if self._reader is not None:
            self._reader.join(timeout=timeout)
        self._raise_if_failed()
        if code != 0:
            raise StereoEncoderError("encoder_failed", f"助手退出码 {code}")
        return self.segments

    def abort(self) -> None:
        process = self._process
        if process is None:
            return
        if process.stdin is not None:
            with _suppress_os_error():
                process.stdin.close()
        if process.poll() is None:
            process.terminate()
            try:
                process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                process.kill()
                with _suppress_os_error():
                    process.wait(timeout=5)
        if self._reader is not None:
            self._reader.join(timeout=5)

    def _raise_if_failed(self) -> None:
        with self._lock:
            failure = self._failure
        if failure is not None:
            raise failure

    def _read_events(self) -> None:
        process = self._process
        assert process is not None and process.stdout is not None
        try:
            for line in process.stdout:
                self._handle_event(line)
        except OSError as error:
            with self._lock:
                self._failure = self._failure or StereoEncoderError(
                    "encoder_failed", f"助手事件流中断：{error}"
                )
        finally:
            self._ready.set()
            self._done.set()

    def _handle_event(self, line: bytes) -> None:
        text = line.strip()
        if not text:
            return
        try:
            event = json.loads(text)
        except ValueError:
            with self._lock:
                self._failure = self._failure or StereoEncoderError(
                    "encoder_failed", "助手输出不是 JSON"
                )
            return
        kind = event.get("event")
        if kind == "ready":
            self._ready.set()
        elif kind == "segment":
            segment = ClosedSegment(
                index=int(event["index"]),
                start_frame=int(event["start_frame"]),
                end_frame=int(event["end_frame"]),
                left_path=str(event["left"]["path"]),
                left_bytes=int(event["left"]["bytes"]),
                right_path=str(event["right"]["path"]),
                right_bytes=int(event["right"]["bytes"]),
            )
            with self._lock:
                self._segments.append(segment)
        elif kind == "done":
            with self._lock:
                self._stats = {
                    key: int(value)
                    for key, value in event.items()
                    if key != "event" and isinstance(value, int)
                }
        elif kind == "error":
            with self._lock:
                self._failure = self._failure or StereoEncoderError(
                    str(event.get("code", "encoder_failed")),
                    str(event.get("message", "助手报告失败")),
                )
            self._ready.set()


class _suppress_os_error:  # noqa: N801 - 与 contextlib.suppress 同用途的轻量写法
    def __enter__(self) -> None:
        return None

    def __exit__(self, exc_type: object, exc: object, traceback: object) -> bool:
        return isinstance(exc, OSError)


def _writev_all(descriptor: int, chunks: Sequence[bytes]) -> None:
    views = [memoryview(chunk) for chunk in chunks if chunk]
    index = 0
    offset = 0
    while index < len(views):
        current = [views[index][offset:], *views[index + 1 :]]
        written = os.writev(descriptor, current)
        if written <= 0:
            raise BrokenPipeError("encoder pipe wrote zero bytes")
        while index < len(views) and written >= len(views[index]) - offset:
            written -= len(views[index]) - offset
            index += 1
            offset = 0
        if written:
            offset += written


def _session_io_or_none() -> NativeSessionIo | None:
    global _SESSION_IO, _SESSION_IO_UNAVAILABLE
    if _SESSION_IO_UNAVAILABLE:
        return None
    if _SESSION_IO is not None:
        return _SESSION_IO
    with _SESSION_IO_LOCK:
        if _SESSION_IO_UNAVAILABLE:
            return None
        if _SESSION_IO is not None:
            return _SESSION_IO
        try:
            _SESSION_IO = create_native_session_io()
        except NativeModuleError:
            _SESSION_IO_UNAVAILABLE = True
            return None
        return _SESSION_IO

File: native/src/stereo_encoder.rs
use crate::session_io;
use serde_json::Value;
use std::collections::BTreeMap;
use std::io::{BufRead, BufReader, Write};
use std::os::fd::AsRawFd;
use std::path::{Path, PathBuf};
use std::process::{Child, ChildStdin, Command, Stdio};
use std::sync::{Arc, Condvar, Mutex};
use std::thread::{self, JoinHandle};
use std::time::{Duration, Instant};

const READY_TIMEOUT: Duration = Duration::from_secs(15);

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct EncoderEventError {
    pub(crate) code: &'static str,
    pub(crate) message: String,
}

impl EncoderEventError {
    fn new(code: &'static str, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct SegmentEvent {
    pub(crate) index: u64,
    pub(crate) start_frame: u64,
    pub(crate) end_frame: u64,
    pub(crate) left_path: String,
    pub(crate) left_bytes: u64,
    pub(crate) right_path: String,
    pub(crate) right_bytes: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) enum EncoderEvent {
    Ready,
    Segment(SegmentEvent),
    Done(BTreeMap<String, i64>),
    Error { code: String, message: String },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct EncoderProcessError {
    pub(crate) code: String,
    pub(crate) message: String,
}

impl EncoderProcessError {
    fn new(code: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            code: code.into(),
            message: message.into(),
        }
    }
}

impl From<EncoderEventError> for EncoderProcessError {
    fn from(error: EncoderEventError) -> Self {
        Self::new(error.code, error.message)
    }
}

impl From<session_io::SessionIoError> for EncoderProcessError {
    fn from(error: session_io::SessionIoError) -> Self {
        Self::new(error.code, error.message)
    }
}

struct ProcessState {
    ready: bool,
    closed: bool,
    failure: Option<EncoderProcessError>,
    segments: Vec<SegmentEvent>,
    stats: BTreeMap<String, i64>,
}

impl ProcessState {
    fn new() -> Self {
        Self {
            ready: false,
            closed: false,
            failure: None,
            segments: Vec::new(),
            stats: BTreeMap::new(),
        }
    }
}

struct SharedProcess {
    state: Mutex<ProcessState>,
    changed: Condvar,
}

pub(crate) struct EncoderProcess {
    executable: PathBuf,
    args: Vec<String>,
    shared: Arc<SharedProcess>,
    child: Option<Child>,
    stdin: Option<ChildStdin>,
    reader: Option<JoinHandle<()>>,
    submitted: u64,
}

impl EncoderProcess {
    #[allow(clippy::too_many_arguments)]
    pub(crate) fn new(
        out_dir: &Path,
        executable: &Path,
        width: u64,
        height: u64,
        fps: u64,
        bitrate_kbps: u64,
        segment_frames: u64,
        path_prefix: &str,
    ) -> Result<Self, EncoderProcessError> {
        if width == 0
            || width % 4 != 0
            || height == 0
            || fps == 0
            || bitrate_kbps == 0
            || segment_frames == 0
        {
            return Err(EncoderProcessError::new(
                "invalid_argument",
                "边录边编码参数无效",
            ));
        }
        Ok(Self {
            executable: executable.to_path_buf(),
            args: vec![
                "--out-dir".to_owned(),
                out_dir.to_string_lossy().into_owned(),
                "--path-prefix".to_owned(),
                path_prefix.to_owned(),
                "--width".to_owned(),
                width.to_string(),
                "--height".to_owned(),
                height.to_string(),
                "--fps".to_owned(),
                fps.to_string(),
                "--bitrate-kbps".to_owned(),
                bitrate_kbps.to_string(),
                "--segment-frames".to_owned(),
                segment_frames.to_string(),
            ],
            shared: Arc::new(SharedProcess {
                state: Mutex::new(ProcessState::new()),
                changed: Condvar::new(),
            }),
            child: None,
            stdin: None,
            reader: None,
            submitted: 0,
        })
    }

    pub(crate) fn start(&mut self) -> Result<(), EncoderProcessError> {
        if self.child.is_some() {
            return Err(EncoderProcessError::new(
                "invalid_state",
                "助手进程只能启动一次",
            ));
        }
        {
            let mut state = self.shared.state.lock().map_err(|_| {
                EncoderProcessError::new("encoder_failed", "encoder process mutex is poisoned")
            })?;
            *state = ProcessState::new();
        }
        let mut child = Command::new(&self.executable)
            .args(&self.args)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .spawn()
            .map_err(|error| {
                EncoderProcessError::new("encoder_unavailable", format!("无法启动助手：{error}"))
            })?;
        let stdout = child
            .stdout
            .take()
            .ok_or_else(|| EncoderProcessError::new("encoder_failed", "无法打开助手事件输出"))?;
        self.stdin = child.stdin.take();
        let shared = Arc::clone(&self.shared);
        self.reader = Some(thread::spawn(move || {
            read_events(stdout, shared);
        }));
        self.child = Some(child);
        if self.stdin.is_none() {
            self.abort();
            return Err(EncoderProcessError::new(
                "encoder_failed",
                "无法打开助手输入",
            ));
        }
        self.wait_ready()
    }

    pub(crate) fn submit(&mut self, jpeg: &[u8]) -> Result<u64, EncoderProcessError> {
        self.raise_if_failed()?;
        let stdin = self
            .stdin
            .as_mut()
            .ok_or_else(|| EncoderProcessError::new("invalid_state", "助手进程未启动"))?;
        let written = session_io::write_encoder_frame(stdin.as_raw_fd(), jpeg)?;
        self.submitted = self.submitted.checked_add(1).ok_or_else(|| {
            EncoderProcessError::new(
                "counter_overflow",
                "encoder submitted frame count overflowed",
            )
        })?;
        Ok(written)
    }

    pub(crate) fn finish(
        &mut self,
        timeout: Duration,
    ) -> Result<Vec<SegmentEvent>, EncoderProcessError> {
        if timeout.is_zero() {
            return Err(EncoderProcessError::new(
                "invalid_argument",
                "finish timeout must be positive",
            ));
        }
        if self.child.is_none() {
            return Err(EncoderProcessError::new("invalid_state", "助手进程未启动"));
        }
        if let Some(mut stdin) = self.stdin.take() {
            let _ = session_io::write_encoder_frame(stdin.as_raw_fd(), &[]);
            let _ = stdin.flush();
        }
        let status = {
            let child = self.child.as_mut().expect("child checked above");
            wait_child(child, timeout).map_err(|error| {
                EncoderProcessError::new("encoder_failed", format!("助手等待失败：{error}"))
            })?
        };
        let Some(status) = status else {
            self.abort();
            return Err(EncoderProcessError::new(
                "encoder_failed",
                "助手未在期限内退出",
            ));
        };
        self.join_reader()?;
        self.raise_if_failed()?;
        if !status.success() {
            return Err(EncoderProcessError::new(
                "encoder_failed",
                format!("助手退出码 {}", status.code().unwrap_or(-1)),
            ));
        }
        Ok(self.segments())
    }

    pub(crate) fn abort(&mut self) {
        let _ = self.stdin.take();
        if let Some(child) = self.child.as_mut() {
            if matches!(child.try_wait(), Ok(None)) {
                let _ = child.kill();
            }
            let _ = child.wait();
        }
        let _ = self.join_reader();
    }

    pub(crate) fn segments(&self) -> Vec<SegmentEvent> {
        match self.shared.state.lock() {
            Ok(state) => state.segments.clone(),
            Err(_) => Vec::new(),
        }
    }

    pub(crate) fn stats(&self) -> BTreeMap<String, i64> {
        match self.shared.state.lock() {
            Ok(state) => state.stats.clone(),
            Err(_) => BTreeMap::new(),
        }
    }

    pub(crate) fn submitted_frames(&self) -> u64 {
        self.submitted
    }

    fn wait_ready(&mut self) -> Result<(), EncoderProcessError> {
        let deadline = Instant::now() + READY_TIMEOUT;
        let mut state = self.shared.state.lock().map_err(|_| {
            EncoderProcessError::new("encoder_failed", "encoder process mutex is poisoned")
        })?;
        while !state.ready && !state.closed && state.failure.is_none() {
            let now = Instant::now();
            if now >= deadline {
                break;
            }
            let timeout = deadline.saturating_duration_since(now);
            let (next_state, wait_result) = self
                .shared
                .changed
                .wait_timeout(state, timeout)
                .map_err(|_| {
                    EncoderProcessError::new("encoder_failed", "encoder process mutex is poisoned")
                })?;
            state = next_state;
            if wait_result.timed_out() {
                break;
            }
        }
        if let Some(error) = state.failure.clone() {
            drop(state);
            self.abort();
            return Err(error);
        }
        if !state.ready {
            let closed = state.closed;
            drop(state);
            self.abort();
            return Err(EncoderProcessError::new(
                "encoder_unavailable",
                if closed {
                    "助手未就绪就退出"
                } else {
                    "助手未在期限内就绪"
                },
            ));
        }
        Ok(())
    }

    fn raise_if_failed(&self) -> Result<(), EncoderProcessError> {
        let state = self.shared.state.lock().map_err(|_| {
            EncoderProcessError::new("encoder_failed", "encoder process mutex is poisoned")
        })?;
        if let Some(error) = state.failure.clone() {
            return Err(error);
        }
        Ok(())
    }

    fn join_reader(&mut self) -> Result<(), EncoderProcessError> {
        if let Some(reader) = self.reader.take() {
            if reader.join().is_err() {
                return Err(EncoderProcessError::new(
                    "encoder_failed",
                    "助手事件线程崩溃",
                ));
            }
        }
        Ok(())
    }
}

impl Drop for EncoderProcess {
    fn drop(&mut self) {
        self.abort();
    }
}

fn read_events(stdout: std::process::ChildStdout, shared: Arc<SharedProcess>) {
    let reader = BufReader::new(stdout);
    for line in reader.split(b'\n') {
        match line {
            Ok(line) => {
                if let Err(error) = handle_event(&shared, &line) {
                    set_failure(&shared, error);
                    break;
                }
            }
            Err(error) => {
                set_failure(
                    &shared,
                    EncoderProcessError::new("encoder_failed", format!("助手事件流中断：{error}")),
                );
                break;
            }
        }
    }
    if let Ok(mut state) = shared.state.lock() {
        state.closed = true;
        shared.changed.notify_all();
    }
}

fn handle_event(shared: &SharedProcess, line: &[u8]) -> Result<(), EncoderProcessError> {
    let Some(event) = parse_event(line)? else {
        return Ok(());
    };
    let mut state = shared.state.lock().map_err(|_| {
        EncoderProcessError::new("encoder_failed", "encoder process mutex is poisoned")
    })?;
    match event {
        EncoderEvent::Ready => {
            state.ready = true;
        }
        EncoderEvent::Segment(segment) => {
            state.segments.push(segment);
        }
        EncoderEvent::Done(stats) => {
            state.stats = stats;
        }
        EncoderEvent::Error { code, message } => {
            state.failure = Some(EncoderProcessError::new(code, message));
        }
    }
    shared.changed.notify_all();
    Ok(())
}

fn set_failure(shared: &SharedProcess, error: EncoderProcessError) {
    if let Ok(mut state) = shared.state.lock() {
        if state.failure.is_none() {
            state.failure = Some(error);
        }
        shared.changed.notify_all();
    }
}

fn wait_child(
    child: &mut Child,
    timeout: Duration,
) -> std::io::Result<Option<std::process::ExitStatus>> {
    let deadline = Instant::now() + timeout;
    loop {
        if let Some(status) = child.try_wait()? {
            return Ok(Some(status));
        }
        if Instant::now() >= deadline {
            return Ok(None);
        }
        thread::sleep(Duration::from_millis(10));
    }
}

pub(crate) fn parse_event(line: &[u8]) -> Result<Option<EncoderEvent>, EncoderEventError> {
    let text = std::str::from_utf8(line)
        .map_err(|_| EncoderEventError::new("encoder_failed", "助手输出不是 UTF-8"))?
        .trim();
    if text.is_empty() {
        return Ok(None);
    }
    let value: Value = serde_json::from_str(text)
        .map_err(|_| EncoderEventError::new("encoder_failed", "助手输出不是 JSON"))?;
    let object = value
        .as_object()
        .ok_or_else(|| EncoderEventError::new("encoder_failed", "助手输出 JSON 不是对象"))?;
    let Some(kind) = object.get("event").and_then(Value::as_str) else {
        return Ok(None);
    };
    match kind {
        "ready" => Ok(Some(EncoderEvent::Ready)),
        "segment" => Ok(Some(EncoderEvent::Segment(parse_segment(object)?))),
        "done" => {
            let stats = object
                .iter()
                .filter_map(|(key, value)| {
                    if key == "event" {
                        None
                    } else {
                        value.as_i64().map(|number| (key.clone(), number))
                    }
                })
                .collect();
            Ok(Some(EncoderEvent::Done(stats)))
        }
        "error" => Ok(Some(EncoderEvent::Error {
            code: object
                .get("code")
                .and_then(Value::as_str)
                .unwrap_or("encoder_failed")
                .to_owned(),
            message: object
                .get("message")
                .and_then(Value::as_str)
                .unwrap_or("助手报告失败")
                .to_owned(),
        })),
        _ => Ok(None),
    }
}

fn parse_segment(
    object: &serde_json::Map<String, Value>,
) -> Result<SegmentEvent, EncoderEventError> {
    Ok(SegmentEvent {
        index: required_u64(object, "index")?,
        start_frame: required_u64(object, "start_frame")?,
        end_frame: required_u64(object, "end_frame")?,
        left_path: required_artifact_path(object, "left")?,
        left_bytes: required_artifact_bytes(object, "left")?,
        right_path: required_artifact_path(object, "right")?,
        right_bytes: required_artifact_bytes(object, "right")?,
    })
}

fn required_u64(
    object: &serde_json::Map<String, Value>,
    key: &'static str,
) -> Result<u64, EncoderEventError> {
    object
        .get(key)
        .and_then(Value::as_u64)
        .ok_or_else(|| EncoderEventError::new("encoder_failed", format!("助手 segment 缺少 {key}")))
}

fn required_artifact_path(
    object: &serde_json::Map<String, Value>,
    key: &'static str,
) -> Result<String, EncoderEventError> {
    object
        .get(key)
        .and_then(Value::as_object)
        .and_then(|artifact| artifact.get("path"))
        .and_then(Value::as_str)
        .map(str::to_owned)
        .ok_or_else(|| {
            EncoderEventError::new("encoder_failed", format!("助手 segment 缺少 {key}.path"))
        })
}

fn required_artifact_bytes(
    object: &serde_json::Map<String, Value>,
    key: &'static str,
) -> Result<u64, EncoderEventError> {
    object
        .get(key)
        .and_then(Value::as_object)
        .and_then(|artifact| artifact.get("bytes"))
        .and_then(Value::as_u64)
        .ok_or_else(|| {
            EncoderEventError::new("encoder_failed", format!("助手 segment 缺少 {key}.bytes"))
        })
}

#[cfg(test)]
mod tests {
    use super::{EncoderEvent, SegmentEvent, parse_event};
    use std::collections::BTreeMap;

    #[test]
    fn parses_ready_segment_done_and_error_events() {
        assert_eq!(
            parse_event(br#"{"event":"ready"}"#).unwrap(),
            Some(EncoderEvent::Ready)
        );
        assert_eq!(
            parse_event(
                br#"{"event":"segment","index":2,"start_frame":10,"end_frame":20,"left":{"path":"video/left_00002.mp4","bytes":123},"right":{"path":"video/right_00002.mp4","bytes":456}}"#,
            )
            .unwrap(),
            Some(EncoderEvent::Segment(SegmentEvent {
                index: 2,
                start_frame: 10,
                end_frame: 20,
                left_path: "video/left_00002.mp4".to_owned(),
                left_bytes: 123,
                right_path: "video/right_00002.mp4".to_owned(),
                right_bytes: 456,
            }))
        );
        assert_eq!(
            parse_event(br#"{"event":"done","frames":7,"ignored":"x"}"#).unwrap(),
            Some(EncoderEvent::Done(BTreeMap::from([(
                "frames".to_owned(),
                7
            )])))
        );
        assert_eq!(
            parse_event(br#"{"event":"error","code":"vpu_failed","message":"bad"}"#).unwrap(),
            Some(EncoderEvent::Error {
                code: "vpu_failed".to_owned(),
                message: "bad".to_owned(),
            })
        );
    }

    #[test]
    fn ignores_empty_and_unknown_events_but_rejects_malformed_json() {
        assert_eq!(parse_event(b"\n").unwrap(), None);
        assert_eq!(parse_event(br#"{"event":"progress"}"#).unwrap(), None);
        let error = parse_event(b"not-json").unwrap_err();
        assert_eq!(error.code, "encoder_failed");
        assert_eq!(error.message, "助手输出不是 JSON");
    }
}
